### PR TITLE
Added support for bundling protobuf as part of the install.

### PR DESCRIPTION
**Pull request by Ferroin**
##### Summary

This adds basic support for bundling a statically linked version of libprotobuf as part of the regular install process. This is currently used by the Prometheus remote write exporter, and will be used in the near future for a new version of the protocol being used to communicate with Netdata Cloud.

If both Netdata Cloud _and_ the Prometheus remote write exporter are disabled, then we will completely skip bundling of protobuf.

The user can pass the option `--with-system-protobuf` to the installer to use a system-provided copy of protobuf. If doing so, we will skip bundling of protobuf, but will print a deferred error at the end of the install warning about how fragile such a setup is.

##### Component Name

area/packaging

##### Test Plan

The build itself seems to work, and from inspection of the resulting binary it looks like it properly includes libprotobuf. Ideally the Prometheus remote writ exporter should be tested with this, though I lack the setup required to do that properly.

##### Additional Information

This adds a non-trivial amount of extra time to the build process, largely due to how big the `protoc` command is (and the fact that it is unconditionally built with support for all languages for which there are protobuf bindings in the official repo). There is unfortunately not much that can be done about this (it’s already set to parallelize the build).

### CODE REVIEW

**Comment by stelfrag** [comment]
Tried on a RPI :face_with_head_bandage: 
```
/usr/bin/ld: /home/pi/netdata/externaldeps/protobuf/src/.libs/libprotobuf.a(arena.o): in function `google::protobuf::internal::ThreadSafeArena::Init(bool)':
arena.cc:(.text+0xaa4): undefined reference to `__atomic_fetch_add_8'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:4831: netdata] Error 1
make[2]: Leaving directory '/home/pi/netdata'
make[1]: *** [Makefile:6013: all-recursive] Error 1
make[1]: Leaving directory '/home/pi/netdata'
make: *** [Makefile:3285: all] Error 2
 FAILED   
```

File: packaging/installer/install-required-packages.sh
Comment (underhood) [inline packaging/installer/install-required-packages.sh:727]: should we add protobuf as well?

**Comment by underhood** [comment]
On FC34 seems to work fine:
<details>
<summary>ldd ./netdata (correctly reports no protobuf)</summary>
```
    linux-vdso.so.1 (0x00007ffca0bc4000)
        libelf.so.1 => /lib64/libelf.so.1 (0x00007fa30784d000)
        libz.so.1 => /lib64/libz.so.1 (0x00007fa307833000)
        libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00007fa307545000)
        libssl.so.1.1 => /lib64/libssl.so.1.1 (0x00007fa3074a9000)
        libuuid.so.1 => /lib64/libuuid.so.1 (0x00007fa3074a0000)
        libuv.so.1 => /lib64/libuv.so.1 (0x00007fa30746b000)
        liblz4.so.1 => /lib64/liblz4.so.1 (0x00007fa307445000)
        libJudy.so.1 => /lib64/libJudy.so.1 (0x00007fa3073f4000)
        libjson-c.so.5 => /lib64/libjson-c.so.5 (0x00007fa3073e1000)
        libsnappy.so.1 => /lib64/libsnappy.so.1 (0x00007fa3073d4000)
        libmongoc-1.0.so.0 => /lib64/libmongoc-1.0.so.0 (0x00007fa307317000)
        libbson-1.0.so.0 => /lib64/libbson-1.0.so.0 (0x00007fa3072d8000)
        libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007fa3070b7000)
        libm.so.6 => /lib64/libm.so.6 (0x00007fa306f73000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fa306f58000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fa306f37000)
        libc.so.6 => /lib64/libc.so.6 (0x00007fa306d68000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fa30788c000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007fa306d61000)
        libsasl2.so.3 => /lib64/libsasl2.so.3 (0x00007fa306d3f000)
        librt.so.1 => /lib64/librt.so.1 (0x00007fa306d34000)
        libresolv.so.2 => /lib64/libresolv.so.2 (0x00007fa306d1a000)
        libzstd.so.1 => /lib64/libzstd.so.1 (0x00007fa306c24000)
        libicuuc.so.67 => /lib64/libicuuc.so.67 (0x00007fa306a39000)
        libmongocrypt.so.0 => /lib64/libmongocrypt.so.0 (0x00007fa306a0b000)
        libcrypt.so.2 => /lib64/libcrypt.so.2 (0x00007fa3069cf000)
        libgssapi_krb5.so.2 => /lib64/libgssapi_krb5.so.2 (0x00007fa306978000)
        libkrb5.so.3 => /lib64/libkrb5.so.3 (0x00007fa30689a000)
        libk5crypto.so.3 => /lib64/libk5crypto.so.3 (0x00007fa306882000)
        libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007fa30687b000)
        libicudata.so.67 => /lib64/libicudata.so.67 (0x00007fa304d60000)
        libkrb5support.so.0 => /lib64/libkrb5support.so.0 (0x00007fa304d4f000)
        libkeyutils.so.1 => /lib64/libkeyutils.so.1 (0x00007fa304d48000)
        libselinux.so.1 => /lib64/libselinux.so.1 (0x00007fa304d1c000)
        libpcre2-8.so.0 => /lib64/libpcre2-8.so.0 (0x00007fa304c85000)
```
</details>
<details>
 <summary>./netdata -W buildinfo (aclk ng built - can't be done without protobuf)</summary>
```
Version: netdata v1.31.0-146-ga4d5ec69f
Configure options:  '--prefix=/home/timo/projects/netdata/netdata-dist/netdata/usr' '--sysconfdir=/home/timo/projects/netdata/netdata-dist/netdata/etc' '--localstatedir=/home/timo/projects/netdata/netdata-dist/netdata/var' '--libexecdir=/home/timo/projects/netdata/netdata-dist/netdata/usr/libexec' '--libdir=/home/timo/projects/netdata/netdata-dist/netdata/usr/lib' '--with-zlib' '--with-math' '--with-user=netdata' '--disable-lto' '--with-bundled-lws' '--with-bundled-protobuf' 'CFLAGS=-Og -g -ggdb -pg -DACLK_SSL_ALLOW_SELF_SIGNED -DNETDATA_INTERNAL_CHECKS=1 -Wall -Wextra -Wformat-signedness -fstack-protector-all -DNETDATA_VERIFY_LOCKS=1 -DACLK_NEWARCH_DEVMODE' 'LDFLAGS=' 'PKG_CONFIG_PATH=/usr/local/lib64/pkgconfig'
Features:
    dbengine:                YES
    Native HTTPS:            YES
    Netdata Cloud:           YES 
    ACLK Next Generation:    YES
    ACLK Legacy:             YES
    TLS Host Verification:   YES
```
</details>

Agent connects to cloud using new protobuf messages after applying this installer patch on New Cloud branch.

**Comment by MrZammler** [comment]
On gentoo it seems to build ok too, same as Timo mentioned for Fedora.

**Comment by MrZammler** [comment]
> Tried on a RPI face_with_head_bandage
> 
> ```
> /usr/bin/ld: /home/pi/netdata/externaldeps/protobuf/src/.libs/libprotobuf.a(arena.o): in function `google::protobuf::internal::ThreadSafeArena::Init(bool)':
> arena.cc:(.text+0xaa4): undefined reference to `__atomic_fetch_add_8'
> collect2: error: ld returned 1 exit status
> make[2]: *** [Makefile:4831: netdata] Error 1
> make[2]: Leaving directory '/home/pi/netdata'
> make[1]: *** [Makefile:6013: all-recursive] Error 1
> make[1]: Leaving directory '/home/pi/netdata'
> make: *** [Makefile:3285: all] Error 2
>  FAILED   
> ```

Building standalone protobuf works here on rpi4, Debian 11, gcc 10 (same configure options as the installer). Will try the installer too.

**Comment by MrZammler** [comment]
We should also maybe check the parameter `--with-system-protobuf` (which appears to be `--use-system-protobuf`), to match the parameters from https://github.com/netdata/netdata/pull/11335 used in configure ?

The scenario is if I build first with bundled protobuf, but then use `--use-system-protobuf` the installer will skip downloading and building protobuf, but the final binary will be linked with the bundled one from previous one (unless I remove `externaldeps/protobuf/`).

File: packaging/installer/install-required-packages.sh
Comment (Ferroin) [reply packaging/installer/install-required-packages.sh:727]: Not if we are planning on bundling them.

I do need to add them as dependencies for the DEB/RPM packages though.

**Comment by Ferroin** [comment]
> We should also maybe check the parameter `--with-system-protobuf` (which appears to be `--use-system-protobuf`), to match the parameters from #11335 used in configure ?
> 
> The scenario is if I build first with bundled protobuf, but then use `--use-system-protobuf` the installer will skip downloading and building protobuf, but the final binary will be linked with the bundled one from previous one (unless I remove `externaldeps/protobuf/`).

That’s only an issue if installing from a git checkout. For the normal case of running the kickstart script, it will switch things over correctly because the source directory gets recreated each time you do an install.

**Comment by MrZammler** [comment]
> > Tried on a RPI face_with_head_bandage
> > ```
> > /usr/bin/ld: /home/pi/netdata/externaldeps/protobuf/src/.libs/libprotobuf.a(arena.o): in function `google::protobuf::internal::ThreadSafeArena::Init(bool)':
> > arena.cc:(.text+0xaa4): undefined reference to `__atomic_fetch_add_8'
> > collect2: error: ld returned 1 exit status
> > make[2]: *** [Makefile:4831: netdata] Error 1
> > make[2]: Leaving directory '/home/pi/netdata'
> > make[1]: *** [Makefile:6013: all-recursive] Error 1
> > make[1]: Leaving directory '/home/pi/netdata'
> > make: *** [Makefile:3285: all] Error 2
> >  FAILED   
> > ```
> 
> Building standalone protobuf works here on rpi4, Debian 11, gcc 10 (same configure options as the installer). Will try the installer too.

It builds ok from the installer too, so maybe a problem with gcc version, or something similar? Also noticed that you're not in a `/tmp` directory (the installer should use one) ?

```
  CXXLD    libprotobuf-lite.la
  CXXLD    libprotobuf.la
  CXXLD    libprotoc.la
  CXXLD    protoc
make[2]: Leaving directory '/tmp/netdata-protobuf-690z34/protobuf-3.17.3/src'
make[1]: Leaving directory '/tmp/netdata-protobuf-690z34/protobuf-3.17.3'
 OK   
```

**Comment by stelfrag** [comment]
That error was on Raspbian with gcc version 8.3.0 (Raspbian 8.3.0-6+rpi1) 

it built protoc 

```
externaldeps/protobuf/src/protoc  --version
libprotoc 3.17.3
```

**Comment by stelfrag** [comment]
> That error was on Raspbian with gcc version 8.3.0 (Raspbian 8.3.0-6+rpi1)
> 
> it built protoc
> 
> ```
> externaldeps/protobuf/src/protoc  --version
> libprotoc 3.17.3
> ```

Ok it works if we can determine and linked with `libatomic`

**Review by underhood — APPROVED** [review]
only issue I know of is the RPi stuff @stelfrag mentioned

**Comment by underhood** [comment]
@stelfrag took a look at the RPi problem and is not related to this PR.
#11406 fixes this issue

**Review by stelfrag — DISMISSED** [review]

**Comment by underhood** [comment]
@Ferroin before hitting the merge button please let #11406 to go in as to not break build for RPi 32bit users

**Comment by Ferroin** [comment]
@underhood It looks like there’s some issue with our build system when building from the source archive produced by `make dist`, see: https://github.com/netdata/netdata/pull/11374/checks?check_run_id=3174470819.

**Comment by underhood** [comment]
@Ferroin can you rebase on latest master?

`#error This file was generated by an older version of protoc which is`... are we bundling the compiled C++ code in dist tarball? Seems like protoc used and headers used are different... not sure what exactly is happening. Will try to look into the github CI files what exactly are we doing there

**Comment by Ferroin** [comment]
> @Ferroin can you rebase on latest master?

Done

> `#error This file was generated by an older version of protoc which is`... are we bundling the compiled C++ code in dist tarball? 

As far as I know we shouldn’t be.

**Comment by underhood** [comment]
@Ferroin I think what might be going on (educated guess) is the build step that does make dist is not using clean source directory.

1. e.g. we are building on system X that uses protoc version A to generate the C++ code
1. then we bundle that build proto files into dist tarball (running make dist on not clean directory)
1. then we compile that files on system that has different protobuf version

**Comment by Ferroin** [comment]
> @Ferroin I think what might be going on (educated guess) is the build step that does make dist is not using clean source directory.
> 
> 1. e.g. we are building on system X that uses protoc version A to generate the C++ code
> 2. then we bundle that build proto files into dist tarball (running make dist on not clean directory)
> 3. then we compile that files on system that has different protobuf version

That’s definitely possible. Do we have the compiled protocol definitions listed as dist files in the build system?

**Comment by underhood** [comment]
well we do `nodist_netdata_SOURCES += $(ACLK_NG_PROTO_BUILT_FILES)` specifically.... but we also do:
```
dist_noinst_DATA = \
...
    aclk/aclk-schemas \
...
```
maybe that does not work as expected and will have to list specifically only `.proto` files in `dist_noinst_DATA`

**Review by stelfrag — DISMISSED** [review]

**Comment by underhood** [comment]
@Ferroin CI failures currently are random (e.g. network failure etc.) but i think the yamllint one should be solved? Afterwards I am approving the PR

**Comment by Ferroin** [comment]
> @Ferroin CI failures currently are random (e.g. network failure etc.) but i think the yamllint one should be solved? Afterwards I am approving the PR

The ClearLinux stuff is indeed broken right now due to upstream network problems.

The two failed Ubuntu ones look like transient errors, so I’m re-queuing those to see if things are fixed.

The yamllint thing can be ignore (it0s only complaining about line length, and we’ve just kind of been ignoring those issues in the GHA workflows because they’re not really issues).

**Review by stelfrag — APPROVED** [review]

**Review by underhood — APPROVED** [review]

**Review by thiagoftsm — APPROVED** [review]
I tested with `protobuf 1.3.3` locally  and it works, it also works when I compiled using `protobuf` from Netdata. For both tests I used `gcc 10.3`.

**Pull request merged**
